### PR TITLE
Add dynamic setup of gpu_limit in gke-job-template module

### PR DESCRIPTION
**Pull request by mohitchaurasia91**
Triggered PR test for all blueprints with usage of `gke-node-pool` and `gke-job-template` module completed with OK status.

### Verification 

- Blueprint: gke-a3-megagpu
  - [Test job logs](https://pantheon.corp.google.com/logs/query;query=my-job-3986%0Aresource.type%3D%22k8s_container%22;cursorTimestamp=2024-11-29T12:18:08.306483383Z;startTime=2024-11-29T04:26:32.984Z;endTime=2024-11-30T04:26:32.984Z?referrer=search&e=13803378&mods=monitoring_api_prod&project=hpc-toolkit-dev) from [daily test build](https://pantheon.corp.google.com/cloud-build/builds;region=global/03eb5e5e-774a-4493-95e1-fca57364de06?e=13803378&mods=monitoring_api_prod&project=hpc-toolkit-dev) running on develop branch, `nvidia-smi` is showing stat from 1 gpu (default value set in `gke-job-template` module)
  - [Test job logs](https://pantheon.corp.google.com/logs/query;query=my-job-ff39%0Aresource.type%3D%22k8s_container%22;cursorTimestamp=2024-11-29T17:37:49.684553214Z;startTime=2024-11-29T04:26:32.984Z;endTime=2024-11-30T04:26:32.984Z?referrer=search&e=13803378&mods=monitoring_api_prod&project=hpc-toolkit-dev) from [PR build](https://pantheon.corp.google.com/cloud-build/builds/a6b0a3a0-252a-464c-a0e7-ec822885cdd0?project=hpc-toolkit-dev&e=13803378&mods=monitoring_api_prod) running on feature branch, `nvidia-smi` is showing stat from 8 gpu (using `gke-node-pool` module output to set this up dynamically), user can still override it using var `requested_gpu_per_pod`

 

### Submission Checklist

NOTE: Community submissions can take up to 2 weeks to be reviewed.

Please take the following actions before submitting this pull request.

* Fork your PR branch from the Toolkit "develop" branch (not main)
* Test all changes with pre-commit in a local branch [#](https://goo.gle/hpc-toolkit#development)
* Confirm that "make tests" passes all tests
* Add or modify unit tests to cover code changes
* Ensure that unit test coverage remains above 80%
* Update all applicable documentation
* Follow Cluster Toolkit Contribution guidelines [#](https://goo.gle/hpc-toolkit-contributing)

### CODE REVIEW

File: modules/compute/gke-job-template/main.tf [hunk 1]
```diff
@@ -61,9 +61,18 @@ locals {
...
+  gpu_limit_string = (
+    alltrue(var.has_gpu) ?                  # if nodes in each pool are configured with GPUs
+    var.requested_gpu_per_pod > 0 ?         # if user supplied requested gpu
+    tostring(var.requested_gpu_per_pod) :   # then honor it
+    (                                       # else
+      local.min_allocatable_gpu > 0 ?       # if allocatable gpu was supplied
+      tostring(local.min_allocatable_gpu) : # then honor it
+      null                                  # else gpu limit is unset
+    ) : null                                # all node pools are configured without gpu, so gpu limit is unset
+  )
```
Comment (annuay-google) [inline modules/compute/gke-job-template/main.tf:75]: Appreciate the effort to explain this out with a detailed comment. Can we instead break this expression into meaningful named smaller conditionals, and then hierarchically combine them?

Comment (annuay-google) [reply modules/compute/gke-job-template/main.tf:75]: Rationale: let the code explain itself, this will need us to keep the code and comment in sync

**Review by annuay-google — CHANGES_REQUESTED** [review]
Please add a section/comment on verifying the change- is there a way to see how many GPUs nvidia-smi has requested for e.g.?

**Comment by mohitchaurasia91** [comment]
> Please add a section/comment on verifying the change- is there a way to see how many GPUs nvidia-smi has requested for e.g.?

Updated PR description with logs from test jobs running on develop and feature branch.

**Review by annuay-google — APPROVED** [review]
LGTM

**Pull request merged**